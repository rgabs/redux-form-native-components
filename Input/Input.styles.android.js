export default {
  input: {
    plain: {
      fontSize: 14,
      marginHorizontal: -5
    },
    primary: {
      fontSize: 14,
      padding: 10
    }
  },
  container: {
    plain: {
      borderColor: '#ccc',
      borderWidth: 1,
      borderRadius: 5,
      backgroundColor: '#FFF',
    },
    primary: {
      marginTop: 10,
      backgroundColor: '#ededed'
    }
  }
};
