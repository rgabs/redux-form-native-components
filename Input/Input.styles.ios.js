export default {
  input: {
    plain: {
      fontSize: 14,
      padding: 7,
      height: 47,
    },
    primary: {
      fontSize: 14,
      padding: 10,
      height: 50,
    }
  },
  container: {
    plain: {
      borderColor: '#ccc',
      borderWidth: 1,
      borderRadius: 5,
      backgroundColor: '#FFF',
    },
    primary: {
      marginTop: 10,
      backgroundColor: '#ededed'
    }
  }
};
