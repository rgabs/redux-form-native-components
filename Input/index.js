import React, {PropTypes, Component} from 'react';
import {View, TextInput} from 'react-native';
import noop from 'lodash/noop';
import styles from './Input.styles';
import ErrorTextIndicator from '../ErrorTextIndicator/ErrorTextIndicator.component';

class Input extends Component {
  _inputRef = null
  _setRef = (ref) => {
    this._inputRef = ref; // ref is required to access the target value in case of onblur and onfocus
  }

  componentWillReceiveProps (nextProps) {
    this._inputRef._lastNativeText = nextProps.input && nextProps.input.value;
  }

  // used _lastNativeText because android doesnt pass targetEvent in onBlur and onFocus
  _onBlurHandler = (inputProps) => () => {
    const {onBlur, onInputChange} = inputProps;
    const val = this._inputRef._lastNativeText;
    onBlur(val);
    onInputChange(val);
  }
  _onFocusHandler = (inputProps) => () => {
    const {onFocus, onInputChange} = inputProps;
    const val = this._inputRef._lastNativeText;
    onFocus(val);
    onInputChange(val);
  }
  _onChangeTextHandler = (inputProps) => (val) => {
    const {onChange, onInputChange} = inputProps;
    onChange(val);
    onInputChange(val);
  }

  render () {
    const {meta, input = {}, containerStyle = {}, inputStyles = {}, value, onFocus = noop, onBlur = noop, onChangeText = noop, onInputChange = noop, disabled = false, theme = 'primary', ...extraProps} = this.props;

    const err = !disabled && (meta && meta.touched && !meta.active && meta.error);

    const wrapperStyle = disabled
    ? [styles.container[theme], styles.disabledInput]
    : styles.container[theme];

    const onChange = onChangeText;

    const inputProps = {onChange, onFocus, onBlur, onInputChange, value, ...input};

    return (
      <View>
        <View style={[wrapperStyle, containerStyle]}>
          <TextInput
          {...extraProps}
          value={inputProps.value}
          ref={this._setRef}
          style={[styles.input[theme], inputStyles]}
          onChangeText={this._onChangeTextHandler(inputProps)}
          underlineColorAndroid={'transparent'}
          onFocus={this._onFocusHandler(inputProps)}
          onBlur={this._onBlurHandler(inputProps, input.value)}
          editable={!disabled}
          placeholderTextColor={theme.placeholderTextColor} />
        </View>
        {err && <ErrorTextIndicator text={err}/>}
      </View>
    );
  }
}

Input.propTypes = {
  meta: PropTypes.object,
  input: PropTypes.object,
  inputStyles: PropTypes.object,
  value: PropTypes.any,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onChangeText: PropTypes.func,
  onInputChange: PropTypes.func,
  disabled: PropTypes.bool,
  theme: PropTypes.string,
  containerStyle: PropTypes.object
};

export default Input;
