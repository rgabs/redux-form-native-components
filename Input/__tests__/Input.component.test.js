import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../index.js';
import {TextInput} from 'react-native';
import {shallow} from 'enzyme';

const onInputChangeSpy = jest.fn();
const onChangeTextSpy = jest.fn();
const onFocusSpy = jest.fn();
const onBlurSpy = jest.fn();

describe('FormComponent: Input component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Input/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Input', () => {
    const wrapper = shallow(<Input value={'test value'} onInputChange={onInputChangeSpy} onBlur={onBlurSpy} onFocus={onFocusSpy} onChangeText={onChangeTextSpy} ref/>);
    const instance = wrapper.instance();

    it('input change callback', () => {
      const input = wrapper.find(TextInput);
      expect(input.at(0).props().value).toEqual('test value');
      instance._onChangeTextHandler({
        onInputChange: instance.props.onInputChange,
        onChange: instance.props.onChangeText}
      )('1111');
      expect(onChangeTextSpy).toBeCalled();
      expect(onInputChangeSpy).toBeCalled();
    });

    it('Focus callback', () => {
      instance._inputRef = {};
      instance._onFocusHandler({
        onInputChange: instance.props.onInputChange,
        onFocus: instance.props.onFocus}
      )();
      expect(onFocusSpy).toBeCalled();
      expect(onInputChangeSpy).toBeCalled();
    });

    it('Blur callback', () => {
      instance._inputRef = {};
      instance._onBlurHandler({
        onInputChange: instance.props.onInputChange,
        onBlur: instance.props.onBlur}
      )();
      expect(onBlurSpy).toBeCalled();
      expect(onInputChangeSpy).toBeCalled();
    });
  });
});
