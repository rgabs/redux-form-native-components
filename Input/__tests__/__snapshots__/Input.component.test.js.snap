// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent: Input component renders correctly 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ededed",
          "marginTop": 10,
        },
        Object {},
      ]
    }
  >
    <TextInput
      editable={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholderTextColor={undefined}
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 16,
            "fontWeight": "300",
            "height": 50,
            "padding": 10,
          },
          Object {},
        ]
      }
      underlineColorAndroid="transparent"
      value={undefined}
    />
  </View>
</View>
`;
